graphs/possible_bipartition.py

```diff
@@ -1,12 +1,86 @@
 # Can be used for BFS
-from collections import deque 
+import collections
+from queue import Queue
+
 
 def possible_bipartition(dislikes):
-    """ Will return True or False if the given graph
+    """ 
+    Given a set of N puppies (numbered 0, 1, 2, ..., N - 1), we would like to split
+    them into two groups of any size to use two play areas.
+
+    Some dogs have a history of fighting with specific other dogs and shouldn't 
+    be put into the same play area.
+
+    Formally, if dislikes[i] = [a, b], it means dog i is not allowed to 
+    put in the same group as dog a or dog b.
+
+    Return true if and only if it is possible to split the dogs into
+    two groups where no fighting will occur.
+
+    ### Example 1
+
+    ```
+    Input: dislikes = [ [],
+                [2, 3],
+                [1, 4],
+                [1],
+                [2]
+                ]
+    Output: true
+```
+
+Explanation: group1 [0, 1, 4], group2 [2, 3]
+        group 1     group 2
+            0           2
+            1           3
+            4
+
+    #  all items within the same set should be disjoint
+    #  A bipartite graph can only have  even edge length cycle.
+    #
+
+    Will return True or False if the given graph
         can be bipartitioned without neighboring nodes put
         into the same partition.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(v + E)
+        Space Complexity: O(n)
     """
-    pass
 
+    # each edge should be a different collors
+
+    if len(dislikes) == 0:
+        return True
+
+    graph = {}
+    graph_items = len(dislikes)
+    for node in range(graph_items):
+        edge = dislikes[node]
+        graph[node] = edge
+    print(graph)
+    # BFS -> loop thorugh each item and check if items are enemies
+
+    def bfs(queue, is_visited, node):
+        queue.put((node, 0))
+        while not queue.empty():
+            curr, value = queue.get()
+            if curr not in is_visited:
+                is_visited[curr] = 0
+            # for the neighbours
+            for next_node in graph[curr]:
+                if next_node not in is_visited:
+                    new_visited = value ^ 1
+                    is_visited[next_node] = new_visited
+                    queue.put((next_node, new_visited))
+                else:
+                    # check if it is in the same group
+                    if is_visited[curr] == is_visited[next_node]:
+                        # they are  supposed to be in different groups
+                        return False
+        return True
+
+    queue = Queue()
+    is_visited = {}
+    for node in range(graph_items):
```

Comment:
🤔 It looks like you're doing a lot of extra work here.

Breadth first search should visit each connected node in the graph and therefore place each connected node in a group and check whether or not that is permitted. For an unconnected node - a dog who doesn't dislike any other dogs - it does not matter which group the node/dog is placed in. Thus, you should only need to perform BFS once.

By looping through each node in the graph and running BFS with that node as the start node, you run BFS n times where n is the number of nodes/dogs you have. Instead of repeating this work over again, how can you instead just run BFS exactly once on the connected nodes in the graph.